// roblox-ts/out/TSTransformer/macros/callMacros.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CALL_MACROS = void 0;
const luau_ast_1 = __importDefault(require("@roblox-ts/luau-ast"));
const convertToIndexableExpression_1 = require("../util/convertToIndexableExpression");
const createTruthinessChecks_1 = require("../util/createTruthinessChecks");
const PRIMITIVE_LUAU_TYPES = new Set([
    "nil",
    "boolean",
    "string",
    "number",
    "table",
    "userdata",
    "function",
    "thread",
    "vector",
]);
exports.CALL_MACROS = {
    assert: (state, node, expression, args) => {
        args[0] = (0, createTruthinessChecks_1.createTruthinessChecks)(state, args[0], node.arguments[0]);
        return luau_ast_1.default.call(luau_ast_1.default.globals.assert, args);
    },
    typeOf: (state, node, expression, args) => luau_ast_1.default.call(luau_ast_1.default.globals.typeof, args),
    typeIs: (state, node, expression, args) => {
        const [value, typeStr] = args;
        const typeFunc = luau_ast_1.default.isStringLiteral(typeStr) && PRIMITIVE_LUAU_TYPES.has(typeStr.value)
            ? luau_ast_1.default.globals.type
            : luau_ast_1.default.globals.typeof;
        return luau_ast_1.default.binary(luau_ast_1.default.call(typeFunc, [value]), "==", typeStr);
    },
    classIs: (state, node, expression, args) => {
        const [value, typeStr] = args;
        return luau_ast_1.default.binary(luau_ast_1.default.property((0, convertToIndexableExpression_1.convertToIndexableExpression)(value), "ClassName"), "==", typeStr);
    },
    opcall: (state, node, expression, args) => {
        const successId = luau_ast_1.default.tempId("success");
        const valueOrErrorId = luau_ast_1.default.tempId("valueOrError");
        state.prereq(luau_ast_1.default.create(luau_ast_1.default.SyntaxKind.VariableDeclaration, {
            left: luau_ast_1.default.list.make(successId, valueOrErrorId),
            right: luau_ast_1.default.call(luau_ast_1.default.globals.pcall, args),
        }));
        const successExp = luau_ast_1.default.map([
            [luau_ast_1.default.strings.success, luau_ast_1.default.bool(true)],
            [luau_ast_1.default.strings.value, valueOrErrorId],
        ]);
        const failureExp = luau_ast_1.default.map([
            [luau_ast_1.default.strings.success, luau_ast_1.default.bool(false)],
            [luau_ast_1.default.strings.error, valueOrErrorId],
        ]);
        return luau_ast_1.default.binary(luau_ast_1.default.binary(successId, "and", successExp), "or", failureExp);
    },
    identity: (state, node, expression, args) => args[0],
};
